outlines/models/gemini.py

```diff
@@ -38,64 +42,98 @@ def format_input(self, model_input):
             The input passed by the user.
 
         """
-        raise NotImplementedError(
-            f"The input type {input} is not available with Gemini. The only available types are `str` and `Vision`."
-        )
-
-    @format_input.register(str)
-    def format_str_input(self, model_input: str):
-        """Generate the `messages` argument to pass to the client when the user
-        only passes a prompt.
-
-        """
-        return {"contents": [model_input]}
-
-    @format_input.register(Vision)
-    def format_vision_input(self, model_input: Vision):
-        """Generate the `messages` argument to pass to the client when the user
-        passes a prompt and an image.
-
-        """
-        return {"contents": [model_input.prompt, model_input.image]}
+        if isinstance(model_input, str):
+            return {"contents": [model_input]}
+        elif isinstance(model_input, Vision):
+            return {"contents": [model_input.prompt, model_input.image]}
+        else:
+            raise TypeError(
+                f"The input type {input} is not available with Gemini. The only available types are `str` and `Vision`."
+            )
 
-    @singledispatchmethod
     def format_output_type(self, output_type):
-        raise NotImplementedError
+        # TODO: `int`, `float` and other Python types could be supported via JSON Schema.
+
+        # Unsupported languages
+        if isinstance(output_type, Regex):
+            raise TypeError(
+                "Neither regex-based structured outputs nor the `pattern` keyword in Json Schema are available with Gemini. Use an open source model or dottxt instead."
+            )
+        elif isinstance(output_type, CFG):
+            raise TypeError(
+                "CFG-based structured outputs are not available with Gemini. Use an open source model or dottxt instead."
+            )
+        elif isinstance(output_type, JsonSchema):
+            raise TypeError(
+                "The Gemini SDK does not accept Json Schemas as an input. Pass a Pydantic model, typed dict or dataclass instead."
+            )
+
+        if output_type is None:
+            return {}
+        elif is_dataclass(output_type):  # structured types
+            return self.format_json_output_type(output_type)
+        elif is_typeddict(output_type):
+            return self.format_json_output_type(output_type)
+        elif isinstance(output_type, type(BaseModel)):
+            return self.format_json_output_type(output_type)
+        # json schema as a dict is accepted but the title keyword is not supported ?!
+        # another restriction: the dict cannot be put in a list
+        elif isinstance(output_type, dict):
+            return self.format_json_output_type(output_type)
+        elif isinstance(output_type, EnumMeta):  # Enum types
```

Comment:
We should have `is_enum`, `is_literal` etc functions for every type in a utils module. Not only would this make the code more readable it makes it easier to maintain should something change with Python's type system